base/src/notification-manager/notification-manager.ts

```diff
@@ -9,33 +9,59 @@
   Copyright Contributors to the Zowe Project.
 */
 
+import { ZoweNotification } from './notification';
 
-export class NotificationManager implements MVDHosting.NotificationManagerInterface {
+let urlSet: boolean = false;
+
+export class ZoweNotificationManager implements MVDHosting.ZoweNotificationManagerInterface {
   private notificationCache: any[];
-  private handlers: MVDHosting.NotificationWatcher[];
+  private handlers: MVDHosting.ZoweNotificationWatcher[];
+  private ws: WebSocket;
 
   constructor() {
-    this.notificationCache = new Array<Notification>();
-    this.handlers = new Array<MVDHosting.NotificationWatcher>();
+    this.notificationCache = new Array<ZoweNotification>();
+    this.handlers = new Array<MVDHosting.ZoweNotificationWatcher>();
   }
 
-  push(notification: Notification): void {
-    this.notificationCache.push(notification);
-    for (let i = 0; i < this.handlers.length; i++) {
-      this.handlers[i].handleMessageAdded();
+  _setURL(url: string): void {
+    if (!urlSet) {
+      this.ws = new WebSocket(url);
+      var _this = this;
+      this.ws.onmessage = function(message) {
+        _this.notificationCache.push((JSON.parse(message.data)['notification']) as ZoweNotification)
+        _this.updateHandlers(JSON.parse(message.data));
+      }
+      this.ws.onclose = () => {
+        _this.ws = new WebSocket(url)
+      }
+      this.ws.onerror = () => {
+        _this.ws = new WebSocket(url)
+      }
+      urlSet = true;
     }
+
   }
 
-  pop(): Notification | void {
-    let n = this.notificationCache.pop();
+  updateHandlers(message: any): void {
     for (let i = 0; i < this.handlers.length; i++) {
-      this.handlers[i].handleMessageAdded();
+      this.handlers[i].handleMessageAdded(message, this.notificationCache.length -1);
     }
-    return n;
   }
 
-  getAll(): Notification[] {
-    let copy: Notification[] = this.notificationCache.slice(0);
+  // getURL() {
+  //   return this.url;
+  // }
+
+  push(notification: ZoweNotification): void {
+    this.ws.send(JSON.stringify(notification))
+  }
+
+  removeFromCache(index: number): void{
+    this.notificationCache.splice(index, 1)
+  }
+
+  getAll(): ZoweNotification[] {
+    let copy: ZoweNotification[] = this.notificationCache.slice(0);
```

Comment:
Dont copy, just hand back contents of this.notificationcache, and let people on the other end decide if they want to copy or not.
You could rename it to this.notification, but a get(), getAll(), getAllByCategory etc is also a clue on what is returned.